pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +507,238 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one of nodeName, nodeSelector or allNodes is required"))
 	case 1:
-		// Okay.
 	default:
-		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(entries), "exactly one field must be set, not several"))
+		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(nodeSelectionFields), "these fields are mutually exclusive"))
+	}
+	allErrs = append(allErrs, validateSet(spec.Devices, resource.ResourceSliceMaxDevices, validateDevice,
+		func(device resource.Device) (string, string) {
+			return device.Name, "name"
+		}, fldPath.Child("devices"))...)
+	if oldSpec != nil {
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.Pool, oldSpec.Pool, fldPath.Child("pool"))...)
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.Driver, oldSpec.Driver, fldPath.Child("driver"))...)
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.NodeName, oldSpec.NodeName, fldPath.Child("nodeName"))...)
 	}
 	return allErrs
 }
 
-// ValidateResourceSlice tests if a ResourceSlice update is valid.
-func ValidateResourceSliceUpdate(resourceSlice, oldResourceSlice *resource.ResourceSlice) field.ErrorList {
-	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceSlice.ObjectMeta, &oldResourceSlice.ObjectMeta, field.NewPath("metadata"))
-	allErrs = append(allErrs, ValidateResourceSlice(resourceSlice)...)
-	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(resourceSlice.NodeName, oldResourceSlice.NodeName, field.NewPath("nodeName"))...)
-	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(resourceSlice.DriverName, oldResourceSlice.DriverName, field.NewPath("driverName"))...)
-	return allErrs
-}
-
-// ValidateResourceClaimParameters tests if a ResourceClaimParameters object is valid for creation.
-func ValidateResourceClaimParameters(parameters *resource.ResourceClaimParameters) field.ErrorList {
-	return validateResourceClaimParameters(parameters, false)
-}
-
-func validateResourceClaimParameters(parameters *resource.ResourceClaimParameters, requestStored bool) field.ErrorList {
-	allErrs := corevalidation.ValidateObjectMeta(&parameters.ObjectMeta, true, apimachineryvalidation.NameIsDNSSubdomain, field.NewPath("metadata"))
-	allErrs = append(allErrs, validateResourceClaimParametersRef(parameters.GeneratedFrom, field.NewPath("generatedFrom"))...)
-	allErrs = append(allErrs, validateDriverRequests(parameters.DriverRequests, field.NewPath("driverRequests"), requestStored)...)
+func validateResourcePool(pool resource.ResourcePool, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validatePoolName(pool.Name, fldPath.Child("name"))...)
+	if pool.ResourceSliceCount <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resourceSliceCount"), pool.ResourceSliceCount, "must be greater than zero"))
+	}
```

Comment:
Do we care about negative generation numbers? I guess not?

Reply:
Indeed we don't. It may look odd to start at -1000, but it's not wrong.

Reply:
Sure. Changed.